pkg/genericcache/genericcache.go

```diff
@@ -0,0 +1,89 @@
+package genericcache
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/portworx/sched-ops/k8s/apiextensions"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+type genericCache struct {
+	crdKindToPluralMap map[string]string
+}
+
+// GenericCache is a generic structure which will implement mostly non-k8s resources related information.
+// Individual members get and set need to be provided by the interface.
+type GenericCache interface {
+	// GetCRDKindToPluralMap returns the crdKindToPluralMap from generic cache singleton.
+	GetCRDKindToPluralMap() map[string]string
+
+	// SetCRDKindToPluralMap will set a new CRD's plural in the map
+	SetCRDKindToPluralMap(kind, plural string)
+}
+
+var (
+	genericCacheLock sync.Mutex
+	genericCacheMap  *genericCache
+)
+
+func Instance() *genericCache {
+	if genericCacheMap == nil {
+		CreateCache()
+	}
+	genericCacheLock.Lock()
+	defer genericCacheLock.Unlock()
+	return genericCacheMap
+}
+
+func getCRDKindToGroupMap() map[string]string {
+	kindToGroupMap := make(map[string]string)
+	crdv1List, err := apiextensions.Instance().ListCRDs()
+	if err == nil {
+		for _, crd := range crdv1List.Items {
+			kindToGroupMap[strings.ToLower(crd.Spec.Names.Kind)] = crd.Spec.Names.Plural
+		}
+	} else if apierrors.IsNotFound(err) {
+		// list and register crds via v1beta1 apis
+		crdv1beta1List, err := apiextensions.Instance().ListCRDsV1beta1()
+		if err != nil {
+			logrus.Warnf("unable to list v1beta1 crds: %v", err)
+			return kindToGroupMap
+		}
+		for _, crd := range crdv1beta1List.Items {
+			if len(crd.Spec.Names.Kind) > 0 {
+				kindToGroupMap[strings.ToLower(crd.Spec.Names.Kind)] = crd.Spec.Names.Plural
+			}
+		}
+	} else {
+		return kindToGroupMap
+	}
+	return kindToGroupMap
+}
+
+func CreateCache() error {
+	genericCacheLock.Lock()
+	defer genericCacheLock.Unlock()
+
+	if genericCacheMap != nil {
+		return fmt.Errorf("generic cache has already been initialized")
+	}
+	genericCacheMap = &genericCache{}
+	genericCacheMap.crdKindToPluralMap = getCRDKindToGroupMap()
+	logrus.Debugf("Current crd plural map: %v", genericCacheMap.crdKindToPluralMap)
+
+	return nil
+}
+
+func (g *genericCache) GetCRDKindToPluralMap() map[string]string {
+	return genericCacheMap.crdKindToPluralMap
```

Comment:
This will lead to `concurrent map read write` error.

I don't think we should return this map which is also governed by a lock to the caller. You might get a watch update and try to add an entry into this map while someone is trying to do a get w/o a lock. 

Make a copy of the map under a lock and return the copy.

Reply:
Agreed. Fixed it.